enterprise/internal/licensing/check.go

```diff
@@ -99,42 +98,72 @@ func (l *licenseChecker) Handle(ctx context.Context) error {
 	return nil
 }
 
+// calcDurationSinceLastCalled calculates the duration to wait
+// before running the next license check. It returns 0 if the
+// license check should be run immediately.
+func calcDurationSinceLastCalled(clock glock.Clock) (time.Duration, error) {
+	lastCalledAt, err := store.Get(lastCalledAtStoreKey).String()
+	if err != nil {
+		return 0, err
+	}
+	lastCalledAtTime, err := time.Parse(time.RFC3339, lastCalledAt)
+	if err != nil {
+		return 0, err
+	}
+
+	if lastCalledAtTime.After(clock.Now()) {
+		return 0, errors.New("lastCalledAt cannot be in the future")
+	}
+
+	elapsed := clock.Since(lastCalledAtTime)
+
+	if elapsed > licenseCheckInterval {
+		return 0, nil
+	}
+	return licenseCheckInterval - elapsed, nil
+}
+
 // StartLicenseCheck starts a goroutine that periodically checks
 // license validity from dotcom and stores the result in redis.
 // It re-runs the check if the license key changes.
-func StartLicenseCheck(ctx context.Context, logger log.Logger, siteID string) {
+func StartLicenseCheck(originalCtx context.Context, logger log.Logger, siteID string) {
+
 	if licenseCheckStarted {
 		logger.Info("license check already started")
 		return
 	}
 	licenseCheckStarted = true
 
+	ctxWithCancel, cancel := context.WithCancel(originalCtx)
+
 	// The entire logic is dependent on config so we will
 	// wait for initial config to be loaded as well as
 	// watch for any config changes
 	conf.Watch(func() {
+		// stop previously running routine
+		cancel()
+		ctxWithCancel, cancel = context.WithCancel(originalCtx)
+
 		prevLicenseToken, _ := store.Get(prevLicenseTokenKey).String()
 		licenseToken := hex.EncodeToString(GenerateHashedLicenseKeyAccessToken(conf.Get().LicenseKey))
-		// skip if license key hasn't changed and already running
-		if prevLicenseToken == licenseToken && routine != nil {
-			return
-		}
-
-		// stop previously running routine
-		if routine != nil {
-			routine.Stop()
+		var initialWaitInterval time.Duration = 0
+		if prevLicenseToken == licenseToken {
+			durationSinceLastCalled, _ := calcDurationSinceLastCalled(glock.NewRealClock())
+			initialWaitInterval = durationSinceLastCalled
```

Comment:
```suggestion
			initialWaitInterval, _ = calcDurationSinceLastCalled(glock.NewRealClock())
```

Might as well shorten it to this